packages/SwingSet/src/vats/vattp/vat-vattp.js

```diff
@@ -16,181 +23,255 @@ import { Far } from '@endo/marshal';
-        openCalled = true;
-
-        // transmitter
-        const transmitter = harden({
-          transmit(msg) {
-            // 'msg' will be a string (vats/comms/outbound.js deliverToRemote)
-            E(connection).send(msg);
-          },
-        });
-        // the comms 'addRemote' API is kind of weird because it's written to
-        // deal with cycles, where vattp has a tx/rx pair that need to be
-        // wired to comms's tx/rx pair. Somebody has to go first, so there's
-        // a cycle. TODO: maybe change comms to be easier
-        const receiverReceiver = harden({
-          setReceiver(receiver) {
-            receivers.set(connection, receiver);
-          },
-        });
-        await E(comms).addRemote(name, transmitter, receiverReceiver);
-        await E(comms).addEgress(name, 0, locatorUnum);
-        gotTheirLocatorUnum(E(comms).addIngress(name, 0));
-      },
-      onReceive(connection, msg) {
-        // setReceiver ought to be called before there's any chance of
-        // onReceive being called
-        E(receivers.get(connection)).receive(msg);
+  // QUESTION: Is there a convention around capitalization here?
+  const mailboxKindHandle = provideKindHandle(baggage, 'mailbox');
+  const initMailboxState = name => ({
+    name,
+    outboundHighestAdded: 0,
+    outboundHighestRemoved: 0,
+    inboundHighestDelivered: 0,
+    inboundReceiver: null,
+  });
+  const makeMailbox = defineDurableKindMulti(
+    mailboxKindHandle,
+    initMailboxState,
+    {
+      transmitter: {
+        transmit: ({ state }, msg) => {
+          const num = state.outboundHighestAdded + 1;
+          D(mailboxDevice).add(state.name, num, msg);
+          state.outboundHighestAdded = num;
+        },
       },
-      onClose(connection, ..._args) {
-        receivers.delete(connection);
-        console.warn(`deleting connection is not fully supported in comms`);
+      setReceiver: {
+        setReceiver: ({ state }, receiver) => {
+          assert(!state.inboundReceiver, X`setReceiver is call-once`);
+          assert(receiver, X`receiver must not be empty`);
+          state.inboundReceiver = receiver;
+        },
       },
-      infoMessage(...args) {
-        void E(cons).log('VatTP connection info:', ...args);
+      inbound: {
+        deliverMessages: ({ state }, newMessages) => {
+          // Minimize interactions with durable storage,
+          // assuming state won't change underneath us
+          // and remote methods won't synchronously throw.
+          const { name, inboundReceiver, inboundHighestDelivered } = state;
+          let newHighestDelivered = inboundHighestDelivered;
+          for (const [num, body] of newMessages) {
+            if (num > newHighestDelivered) {
+              // TODO: SO() / sendOnly()
+              E(inboundReceiver).receive(body);
+              newHighestDelivered = num;
+              D(mailboxDevice).ackInbound(name, num);
```

Comment:
Technically this could throw.. there's no good reason for the mailbox device to do so, but we should consider what would happen if it did.

Ok, so say we get three `newMessages` here, 1/2/3. And say the device call throws during message 2, so we've already called `E(inboundReceiver).receive(1)` and `(2)`. Now the device might or might not have implemented its ack update (setting the mailbox inbound `ack` value to 2) before it throws. Once it does throw, this loop would be exited, and the function would exit, so we wouldn't set `state.inboundHighestDelivered=` at the end leaving it at `0`.

The mailbox device that pushed the `deliverMessages` call is not waiting on a result promise, so it won't notice the error. It thinks the `deliverMessages` was reliable, so (assuming the host keeps running) the next call will get message 4. We'll deliver that to the receiver (because the `num` is higher than our remembered `0`), so the receiver will miss out on 3 entirely. We currently have the comms protocol include explicit sequence numbers, which means comms will notice the gap, and `receive()` will throw. Comms will throw on every received message that isn't `3`, effectively breaking that particular connection. If/when we turn off explicit seqnums, comms would just get very confused, which is probably worse.

In some contexts (more likely an ag-solo than a chain), if the host happened to shut down and get restarted after this error, the mailbox device is within its rights to re-deliver all the un-acked inbound messages, especially when it receives an even newer message (4) later. So it might make a new `deliverMessages` call with 2/3 or 2/3/4. Assuming `ackInbound` doesn't throw again, we'd receive and process the `2` (causing a duplicate comms message, which would cause comms-with-explicit-seqnums to throw). vattp doesn't pay attention to the comms results, so vattp will charge ahead. It does `D().ackInbound()` with `2`, which is fine regardless of whether the earlier (throwing) call managed to record or not, because all `ackInbound` does is set a piece of mailbox state (it doesn't insist that the new value is one-larger or at-all-larger than the old value). So then vattp can continue ahead to `3`, which *will* be accepted by comms, and then acked, and everything can get back on track.

In the old version, the device call was the last thing we did before looping (we finished updated our local state just before calling into the device). That design was resistant to making duplicate deliveries to comms (the recorded seqnum would be accurate, so they'd be skipped by `deliverInboundMessages`), but could still experience gaps. So, not rock-solid, but maybe it has one fewer failure mode.

So, on the whole, I think I'd prefer to do the `state.inboundHighestDelivered` assignment within the loop, just before making the device call. The cost is an extra virtual-object property write, but remember that `VIRTUAL_OBJECT_CACHE_SIZE` means that we won't write this back to the DB until either a `delivery.bringOutYourDead` or the object gets pushed out of the LRU cache, so I don't think it will cause extra syscalls or DB traffic, making it effectively free.

Reply:
> Technically this could throw.. there's no good reason for the mailbox device to do so, but we should consider what would happen if it did.
> 
> Ok, so say we get three `newMessages` here, 1/2/3. And say the device call throws during message 2, so we've already called `E(inboundReceiver).receive(1)` and `(2)`. Now the device might or might not have implemented its ack update (setting the mailbox inbound `ack` value to 2) before it throws. Once it does throw, this loop would be exited, and the function would exit, so we wouldn't set `state.inboundHighestDelivered=` at the end leaving it at `0`.

Isn't `D(validDevice).method(…)` guaranteed not to synchronously throw? That was my understanding as commented above[^1], and it seems to align with https://github.com/Agoric/agoric-sdk/blob/0e8644a34f996242935d4a525c631ae32e1da462/packages/SwingSet/src/liveslots/liveslots.js#L1020-L1033 (assuming that is the right definition).

> So, on the whole, I think I'd prefer to do the `state.inboundHighestDelivered` assignment within the loop, just before making the device call. The cost is an extra virtual-object property write, but remember that `VIRTUAL_OBJECT_CACHE_SIZE` means that we won't write this back to the DB until either a `delivery.bringOutYourDead` or the object gets pushed out of the LRU cache, so I don't think it will cause extra syscalls or DB traffic, making it effectively free.

Oh, I didn't realize we already had that. I observed a testing speedup from this change (and in fact that's what motivated it), but I can change it back if you don't find the above analysis of `D` sufficiently compelling.

[^1]: "_Minimize interactions with durable storage, assuming state won't change underneath us and remote methods won't synchronously throw._"

Reply:
Yeah, the vat-vattp code is written with the expectation that the device call won't throw, I'm just thinking deviously this week.

The real code to look at is the `DeviceHandler` definition just above that chunk:
https://github.com/Agoric/agoric-sdk/blob/0e8644a34f996242935d4a525c631ae32e1da462/packages/SwingSet/src/liveslots/liveslots.js#L997-L1015

`D(mailboxDevice)` on a real device node won't throw, but the `.ackInbound(..)` could throw if the args are unserializable, are too big, hold a promise, if the underlying device invocation throws, or if it returns a promise. None of those are things well-behaved code would do, though.

But if you saw a speedup, then yeah, go with the faster version. I don't have any good reason to believe that my ordering would be meaningfully more robust in practice (if the device throws, who knows what else will go wrong).